Initial support for chain operation modes

mpapierski opened the pull request:
Closes CasperLabs/product#54

This PR adds initial support for the private chain operation mode.

Commit:
Hide EngineConfig construction under builder.

This should give greater isolation between EE modifications and effects
on the rest of the system.

Commit:
Added admin control management for private chains

Commit:
Ensure private chain accounts cant manage keys.

Commit:
Extra validation for private chain genesis.

goral09 commented:
@mpapierski can you provide more context in the PR description? What's the request/goal here? What we're trying to achieve, what kind of feature are we adding? How are we going about it? I.e. what got changes/added/removed to support it?

Commit:
Fix CI

Commit:
Fix AS errors

goral09 commented on execution_engine/src/core/engine_state/engine_config.rs:
```diff
@@ -72,4 +57,99 @@ impl EngineConfig {
+    }
+}
+
+/// This is a builder pattern applied to the [`EngineConfig`] structure to shield any changes to the
+/// constructor, or contents of it from the rest of the system.
+///
+/// Any field that isn't specified will be defaulted.
+#[derive(Default, Debug)]
+pub struct EngineConfigBuilder {
+    max_query_depth: Option<u64>,
+    max_associated_keys: Option<u32>,
+    max_runtime_call_stack_height: Option<u32>,
+    wasm_config: Option<WasmConfig>,
+    system_config: Option<SystemConfig>,
+    chain_kind: Option<ChainKind>,
+}
```
Why not add the new field to the `new` constructor? I don't think that using the defaults for struct like this is a good idea, seems like an easy way to make mistakes.

goral09 commented on execution_engine/src/shared/chain_kind.rs:
```diff
@@ -0,0 +1,72 @@
+pub enum ChainKind {
+    /// Public (default) mode
+    Public,
+    /// Private chain mode with extra capabilities only available if this mode is set.
+    Private,
```
Each private chain could choose a different set of "special features". I don't think we should generalize and assume that all of them will want the same thing.

Commit:
Create private chain accounts after transfer.

Accounts created after a transfer to a non-existing account are also
modified to contain new action thresholds, and extra associated keys.

mpapierski commented:
Closed in favor of #2883
